Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp

```diff
@@ -391,1286 +2515,361 @@ double BytecodeInterpreter::read_value<double>(ReadonlyBytes data)
-    case Instructions::i8x16_narrow_i16x8_u.value():
-        return binary_numeric_operation<u128, u128, Operators::VectorNarrow<16, u8>>(configuration);
-    case Instructions::i16x8_narrow_i32x4_s.value():
-        return binary_numeric_operation<u128, u128, Operators::VectorNarrow<8, i16>>(configuration);
-    case Instructions::i16x8_narrow_i32x4_u.value():
-        return binary_numeric_operation<u128, u128, Operators::VectorNarrow<8, u16>>(configuration);
-    case Instructions::i16x8_q15mulr_sat_s.value():
-        return binary_numeric_operation<u128, u128, Operators::VectorIntegerBinaryOp<8, Operators::SaturatingOp<i16, Operators::Q15Mul>, MakeSigned>>(configuration);
-    case Instructions::f32x4_convert_i32x4_s.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<4, 4, u32, i32, Operators::Convert<f32>>>(configuration);
-    case Instructions::f32x4_convert_i32x4_u.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<4, 4, u32, u32, Operators::Convert<f32>>>(configuration);
-    case Instructions::f64x2_convert_low_i32x4_s.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<2, 4, u64, i32, Operators::Convert<f64>>>(configuration);
-    case Instructions::f64x2_convert_low_i32x4_u.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<2, 4, u64, u32, Operators::Convert<f64>>>(configuration);
-    case Instructions::f32x4_demote_f64x2_zero.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<4, 2, u32, f64, Operators::Convert<f32>>>(configuration);
-    case Instructions::f64x2_promote_low_f32x4.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<2, 4, u64, f32, Operators::Convert<f64>>>(configuration);
-    case Instructions::i32x4_trunc_sat_f64x2_s_zero.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<4, 2, u32, f64, Operators::SaturatingTruncate<i32>>>(configuration);
-    case Instructions::i32x4_trunc_sat_f64x2_u_zero.value():
-        return unary_operation<u128, u128, Operators::VectorConvertOp<4, 2, u32, f64, Operators::SaturatingTruncate<u32>>>(configuration);
+
+    for (auto id : forced_stack_values)
+        stack_forced_roots.set(final_roots[id.value()]);
+    for (auto& interval : intervals)
+        interval.forced_to_stack = stack_forced_roots.contains(final_roots[interval.value_id.value()]);
+
+    quick_sort(intervals, [](auto const& a, auto const& b) {
+        return a.start < b.start;
+    });
+
+    HashMap<ValueID, Dispatch::RegisterOrStack> value_alloc;
+    RedBlackTree<size_t, ActiveReg> active_by_end;
+
+    auto expire_old_intervals = [&](IP current_start) {
+        while (true) {
+            auto it = active_by_end.find_smallest_not_below_iterator(current_start.value());
+            if (it.is_end())
+                break;
+            active_by_end.remove(it.key());
+        }
```

Comment:
I have the feeling that RB trees might offer a really fast way of pruning values above a threshold, but I did not take a DSA class

Reply:
I really don't wanna do another AK change, I'll switch this over if we add that kinda API later.